x/ecocredit/server/marketplace/sell_test.go

```diff
@@ -0,0 +1,150 @@
+	}).Times(2)
+
+	sellTime := time.Now()
+	res, err := s.k.Sell(s.ctx, &v1.MsgSell{
+		Owner:  s.addr.String(),
+		Orders: []*v1.MsgSell_Order{
+			{BatchDenom: batchDenom, Quantity: "10",AskPrice: &ask, DisableAutoRetire: false, Expiration: &sellTime},
+			{BatchDenom: batchDenom, Quantity: "10",AskPrice: &ask, DisableAutoRetire: false, Expiration: &sellTime},
+		},
+	})
+	assert.NilError(t, err)
+	assert.Equal(t, 2, len(res.SellOrderIds))
+
+	it, err := s.marketStore.SellOrderStore().List(s.ctx, marketApi.SellOrderSellerIndexKey{})
+	assert.NilError(t, err)
+	count := 0
+	for it.Next() {
+		val, err := it.Value()
+		assert.NilError(t, err)
+		assert.Equal(t, "10", val.Quantity)
+		assert.Equal(t, ask.Amount.String(), val.AskPrice)
+		count++
+	}
+	assert.Equal(t, 2, count)
+}
+
+func TestSell_Invalid(t *testing.T) {
+	t.Parallel()
+	s := setupBase(t)
+	any := gomock.Any()
+	batchDenom := "C01-20200101-20200201-001"
+	start, end := timestamppb.Now(), timestamppb.Now()
+	ask := sdk.NewInt64Coin("ufoo", 10)
+	creditType := ecocredit.CreditType{
+		Name:         "carbon",
+		Abbreviation: "C",
+		Unit:         "tonnes",
+		Precision:    6,
+	}
+	testSellSetup(t, s, batchDenom, ask.Denom, ask.Denom[1:], "C01", start, end, creditType)
+	sellTime := time.Now()
+
+	// invalid batch
+	_, err := s.k.Sell(s.ctx, &v1.MsgSell{
+		Owner:  s.addr.String(),
+		Orders: []*v1.MsgSell_Order{
+			{BatchDenom: "foo-bar-baz-001", Quantity: "10", AskPrice: &ask, DisableAutoRetire: true, Expiration: &sellTime},
+		},
+	})
+	assert.ErrorContains(t, err, "batch denom foo-bar-baz-001")
+
+	s.paramsKeeper.EXPECT().GetParamSet(any, any).Do(func(any interface{}, p *ecocredit.Params) {
+		p.CreditTypes = []*ecocredit.CreditType{&creditType}
+	}).Times(3)
+
+	// market not found
+	invalidDenom := sdk.NewInt64Coin("baz", 50)
+	_, err = s.k.Sell(s.ctx, &v1.MsgSell{
+		Owner:  s.addr.String(),
+		Orders: []*v1.MsgSell_Order{
+			{BatchDenom: batchDenom, Quantity: "10", AskPrice: &invalidDenom, DisableAutoRetire: true, Expiration: &sellTime},
+		},
+	})
+	assert.ErrorContains(t, err, "market: batch denom", batchDenom)
+
+	// invalid balance
+	_, err = s.k.Sell(s.ctx, &v1.MsgSell{
+		Owner:  s.addr.String(),
+		Orders: []*v1.MsgSell_Order{
+			{BatchDenom: batchDenom, Quantity: "10000000000", AskPrice: &ask, DisableAutoRetire: true, Expiration: &sellTime},
+		},
+	})
+	assert.ErrorContains(t, err, ecocredit.ErrInsufficientFunds.Error())
+
+	// order expiration not in the future
+	s.sdkCtx = s.sdkCtx.WithBlockTime(time.Now())
+	s.ctx = sdk.WrapSDKContext(s.sdkCtx)
+	invalidExpirationTime, err := time.Parse("2006-01-02", "1500-01-01")
+	_, err = s.k.Sell(s.ctx, &v1.MsgSell{
+		Owner:  s.addr.String(),
+		Orders: []*v1.MsgSell_Order{
+			{BatchDenom: batchDenom, Quantity: "10", AskPrice: &ask, DisableAutoRetire: true, Expiration: &invalidExpirationTime},
+		},
+	})
+	assert.ErrorContains(t, err, "expiration must be in the future")
+}
```

Comment:
This test excludes a check for whether denom is allowed, which is ok for now, but we should add a todo to add it and link the todo to #624.